src/components/tooltip/CallNode.js

```diff
@@ -57,15 +62,241 @@ type Props = {|
  */
 export class TooltipCallNode extends React.PureComponent<Props> {
-  _renderTimings(
+  _renderCategoryTimings(
+    maybeTimings: ?TimingsForPath,
+    maybeDisplayData: ?CallNodeDisplayData
+  ) {
+    if (!maybeTimings || !maybeDisplayData) {
+      return null;
+    }
+    const { totalTime, selfTime } = maybeTimings.forPath;
+    if (!totalTime.breakdownByCategory) {
+      return null;
+    }
+
+    const totalTimeBreakdownByCategory: BreakdownByCategory =
+      totalTime.breakdownByCategory;
+
+    const { thread, weightType, categories } = this.props;
+
+    // JS Tracer threads have data relevant to the microsecond level.
+    const isHighPrecision = Boolean(thread.isJsTracer);
+
+    const categoriesAndTime: Array<{
+      categoryName: string,
+      category: number,
+      subCategory: number, // -1 for category headers
+      totalTime: number,
+      selfTime: number,
+    }> = [];
+    totalTimeBreakdownByCategory.forEach(
+      ({ entireCategoryValue, subcategoryBreakdown }, category) => {
+        if (entireCategoryValue === 0) {
+          return;
+        }
+        const selfTimeValue = selfTime.breakdownByCategory
+          ? selfTime.breakdownByCategory[category].entireCategoryValue
+          : 0;
+
+        const filteredSubcategoryBreakdown: Array<{
+          subcategory: number,
+          self: number,
+        }> = subcategoryBreakdown
+          .map((subval, i) => {
+            return { subcategory: i, self: subval };
+          })
+          .filter((sub) => sub.self !== 0);
+        if (filteredSubcategoryBreakdown.length === 1) {
+          const { subcategory } = filteredSubcategoryBreakdown[0];
+          const categoryName =
+            categories[category].name +
+            (subcategory !== 0
+              ? ': ' + categories[category].subcategories[subcategory]
+              : '');
+          categoriesAndTime.push({
+            categoryName,
+            category,
+            subCategory: -1,
+            selfTime: selfTimeValue,
+            totalTime: entireCategoryValue,
+          });
+          return;
+        }
+        const categoryName = categories[category].name;
+        categoriesAndTime.push({
+          categoryName,
+          category,
+          subCategory: -1,
+          selfTime: selfTimeValue,
+          totalTime: entireCategoryValue,
+        });
+        if (filteredSubcategoryBreakdown[0].subcategory === 0) {
+          const other = filteredSubcategoryBreakdown[0];
+          filteredSubcategoryBreakdown.pop();
+          filteredSubcategoryBreakdown.push(other);
+        }
+        filteredSubcategoryBreakdown.forEach((sub) => {
+          const { self, subcategory } = sub;
+          const selfTimeValue = selfTime.breakdownByCategory
+            ? selfTime.breakdownByCategory[category].subcategoryBreakdown[
+                subcategory
+              ]
+            : 0;
+          categoriesAndTime.push({
+            categoryName,
```

Comment:
Why don't you push the subcategory name here? Then you don't need `getLastCategoryPartLabel` anymore, you can just use the `categoryName` from this structure...